lambdas/cfn-cognito-resource-server/notify-cfn.js

```diff
@@ -0,0 +1,109 @@
+const crypto = require('crypto')
+const inspect = require('util').inspect
+
+let lastRequestId = ''
+
+/**
+ * 
+ * The config object used to define the response sent to CloudFormation.
+ * 
+ * @typedef {Object} ResponseConfig
+ * 
+ * @property {Object} event - The event object passed in from the lambda function.
+ * @property {Object} context - The context object passed in from the lambda function. Required if logStream is not present.
+ * @property {string} responseStatus - Indicates if the response was a SUCCESS or FAILURE.
+ * @property {Object} responseData - Data to pass to a successful response. Required on SUCCESSes.
+ * @property {string|Object} error - The error object or message that indicates failure. Required on FAILUREs.
+ * @property {string} [physicalResourceId] - An optional id. Generated from the responseData if not provided.
+ * @property {string} [logicalResourceId] - An optional id. Uses the generating lambda function if not provided.
+ */
+
+/**
+ * Uses the cfn-response library to notify CloudFormation that this custom resource is done with the task it was
+ * invoked to do. This could be the response to a create / update request (which would upload files from S3) or a
+ * delete request (which would delete files from S3).
+ * 
+ * Returns a Promise b/c it's being used inside an async function.
+ *
+ * @param {ResponseConfig} config - context lambda function context
+ */
+function notifyCFN ({ event, context, responseStatus, responseData, error, physicalResourceId, logicalResourceId }) {
+  // if (lastRequestId === event.RequestId)
+    // return Promise.reject("Attempted to run `notifyCFN` more than once with the same responseURL. I'm afraid I can't let you do that, Dave.")
+  
+  // lastRequestId = event.RequestId
+
+  return new Promise((resolve, reject) => {
+    let errorMessage = error ? error.message || error.error || error.errorMessage || error : ''
+
+    // if there's no response message, default it based on presence of an error
+    if (!responseStatus)
+      responseStatus = error ? 'FAILED' : 'SUCCESS' 
+
+    // if there's no physicalResourceId and it's a SUCCESS, fake a physicalResourceId from the responseData
+    if (!physicalResourceId)
+      physicalResourceId = crypto.createHash('md5').update(inspect(responseData) || '{}').digest('hex')
+
+    if (!logicalResourceId) logicalResourceId = event.LogicalResourceId
+
+    var responseBody = JSON.stringify({
+      Status: responseStatus,
+      Reason: (errorMessage || '') + "\n For more details, see CloudWatch group (" + context.logGroupName + ') and stream (' + context.logStreamName + ')',
+      PhysicalResourceId: physicalResourceId,
+      StackId: event.StackId,
+      RequestId: event.RequestId,
+      LogicalResourceId: logicalResourceId, // the lambda function; do we want this?
+      Data: { "message": "Success" }
```

Comment:
Flatttttt it, genericize, copy across. Long term fix is probably lambda layer or build-step.